thetis/utility.py

```diff
@@ -306,11 +306,64 @@ def comp_tracer_mass_2d(eta, bath, scalar_func):
     :arg bath: bathymetry :class:`Function`
     :arg scalar_func: scalar :class:`Function` to integrate
     """
-
     val = assemble((eta+bath)*scalar_func*dx)
     return val
 
 
+def comp_tracer_total_mass_2d(var, tracer_name):
+    """
+    Computes total tracer mass in the 2D domain accounting for tracer leaving through
+    boundary conditions and source terms
+    :arg var: class:`DiagnosticCallback`; callback object used to input values and
+                                          record output values
+    :arg tracer_name :class:`string` of function name of interest
+    """
+
+    # read in necessary variables from solver object
+    eta = var.solver_obj.fields.elev_2d.copy(deepcopy=True)
+    vel = var.solver_obj.fields.uv_2d.copy(deepcopy=True)
+
+    scalar_func = var.solver_obj.fields[tracer_name].copy(deepcopy=True)
```

Comment:
You are making copies of the elevation, velocity, and c fields every time this function is being called. There must be a better way.

Reply:
I have changed this so that it is no longer making copies but instead just acting as a pointer and the code still works. Would this be more appropriate?

Reply:
I think so. You are not writing to those fields so it should be fine.